homeassistant/components/netatmo/data_handler.py

```diff
@@ -0,0 +1,174 @@
+"""The Netatmo data handler."""
+import asyncio
+from datetime import timedelta
+from functools import partial
+import logging
+from time import time
+from typing import Dict, List
+
+import pyatmo
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+
+from .const import AUTH, DOMAIN, MANUFACTURER
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DATA_CLASSES = {
+    "WeatherStationData": pyatmo.WeatherStationData,
+    "HomeCoachData": pyatmo.HomeCoachData,
+    "CameraData": pyatmo.CameraData,
+    "HomeData": pyatmo.HomeData,
+    "HomeStatus": pyatmo.HomeStatus,
+    "PublicData": pyatmo.PublicData,
+}
+
+MAX_CALLS_1H = 20
+PARALLEL_CALLS = 3
+DEFAULT_INTERVALS = {
+    "HomeData": 900,
+    "HomeStatus": 300,
+    "CameraData": 900,
+    "WeatherStationData": 300,
+    "HomeCoachData": 300,
+    "PublicData": 600,
+}
+SCAN_INTERVAL = 60
+
+
+class NetatmoDataHandler:
+    """Manages the Netatmo data handling."""
+
+    # Central class to manage the polling data from the Netatmo API
+    # as well as the push data from the webhook
+    #
+    # Create one instance of the handler and store it in hass.data
+    #
+    # Register entities of its platforms when added to HA
+    # to receive signals once new data is available
+    #
+    # Fetch data for all data classes for the first time
+    # to gather all available entities
+    # then only update periodically the registered data classes and
+    # dispatch signals for the registered entities to fetch the new data
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry):
+        """Initialize self."""
+        self.hass = hass
+        self._auth = hass.data[DOMAIN][entry.entry_id][AUTH]
+        self.listeners: List[CALLBACK_TYPE] = []
+        self._data_classes: Dict = {}
+        self.data = {}
+        self._queue: List = []
+        self._webhook: bool = False
+
+        self.lock = asyncio.Lock()
+
+    async def async_setup(self):
+        """Set up the Netatmo data handler."""
+
+        async def async_update(event_time):
+            """Update device."""
+            queue = self._queue[0:PARALLEL_CALLS]
+            for _ in queue:
+                self._queue.append(self._queue.pop(0))
```

Comment:
There should be a cleaner way to do this.  I'll come back to looking at it when I have more time.

Reply:
I switched to using deque and rotate that.